src/range_query/sqrt_decomp.rs

```diff
@@ -129,14 +131,54 @@ impl PiecewiseLinearFn {
     fn update_envelope(&mut self) {
         self.recent_lines.extend(self.sorted_lines.drain(..));
         self.recent_lines.sort_unstable();
-        for (slope, intercept) in self.recent_lines.drain(..) {
-            // TODO: do convex hull trick algorithm
-            self.sorted_lines.push((slope, intercept));
+        self.intersections.clear();
+
+        'outer: for (m1, b1) in self.recent_lines.drain(..) {
+            while let Some(&(m2, b2)) = self.sorted_lines.last() {
+                // The lines are sorted by intercept so an incoming line with
+                // the same slope can never beat the previous line.
+                if m2 == m1 {
+                    continue 'outer;
+                }
+                let new_intersection = (b1 - b2) as f64 / (m2 - m1) as f64;
+                if let Some(&last_intersection) = self.intersections.last() {
+                    if new_intersection < last_intersection {
+                        self.intersections.pop();
+                        self.sorted_lines.pop();
+                        continue;
+                    }
+                }
+                self.intersections.push(new_intersection);
+                break;
+            }
+            self.sorted_lines.push((m1, b1));
+        }
+    }
+
+    fn eval_in_envelope(&self, x: i64) -> i64 {
+        if self.sorted_lines.is_empty() {
+            return i64::MAX;
```

Comment:
This check is no longer needed after adding `unwrap_or`, as commented below.

Reply:
We index into sorted_lines later in the function, so I think the check is still necessary here.

Reply:
Hm ok fair, I don't see a nice way to avoid duplicating the i64::MAX